Mathlib/NumberTheory/ArithmeticFunction.lean

```diff
@@ -133,61 +142,63 @@ end One
 
 end Zero
 
+/-- Coerc an arithmetic function with values in `ℕ` to one with values in `R`. We cannot inline
+this in `natCoe` because it gets unfolded too much. -/
+@[coe]  -- porting note: added `coe` tag.
+def natToArithmeticFunction [AddMonoidWithOne R] :
+  (ArithmeticFunction ℕ) → (ArithmeticFunction R) :=
+  fun f => ⟨fun n => ↑(f n), by simp⟩
+
 instance natCoe [AddMonoidWithOne R] : Coe (ArithmeticFunction ℕ) (ArithmeticFunction R) :=
-  ⟨fun f =>
-    ⟨↑(f : ℕ → ℕ), by
-      trans ↑(f 0)
-      rfl
-      simp⟩⟩
+  ⟨natToArithmeticFunction⟩
 #align nat.arithmetic_function.nat_coe Nat.ArithmeticFunction.natCoe
 
-@[simp]
-theorem natCoe_nat (f : ArithmeticFunction ℕ) : (↑f : ArithmeticFunction ℕ) = f :=
-  ext fun _ => cast_id _
-#align nat.arithmetic_function.nat_coe_nat Nat.ArithmeticFunction.natCoe_nat
+-- Porting note : there was a theorem `(↑f : ArithmeticFunction ℕ) = f` that is now a syntactic
+-- equality.
```

Comment:
Probably, there is some `Coe*` instance for `id` and Lean 4 finds it before `natToArithmeticFunction`. You should restore this lemma with `natToArithmeticFunction f` in the LHS.

Reply:
Added, happy linter!